pkg/tracex/trace.go

```diff
@@ -0,0 +1,241 @@
+	"github.com/ooni/minivpn/internal/optional"
+	"github.com/ooni/minivpn/internal/session"
+)
+
+const (
+	handshakeEventStateChange = iota
+	handshakeEventPacketIn
+	handshakeEventPacketOut
+	handshakeEventPacketDropped
+)
+
+// HandshakeEventType indicates which event we logged.
+type HandshakeEventType int
+
+// Ensure that it implements the Stringer interface.
+var _ fmt.Stringer = HandshakeEventType(0)
+
+// String implements fmt.Stringer
+func (e HandshakeEventType) String() string {
+	switch e {
+	case handshakeEventStateChange:
+		return "state"
+	case handshakeEventPacketIn:
+		return "packet_in"
+	case handshakeEventPacketOut:
+		return "packet_out"
+	case handshakeEventPacketDropped:
+		return "packet_dropped"
+	default:
+		return "unknown"
+	}
+}
+
+// event is one handshake event collected by this [model.HandshakeTracer].
+type event struct {
+	// EventType is the type for this event.
+	EventType HandshakeEventType
+
+	// Stage is the stage of the handshake negotiation we're in.
+	Stage session.SessionNegotiationState
+
+	// AtTime is the time for this event, relative to the start time.
+	AtTime time.Duration
+
+	// Tags is an array of tags that can be useful to interpret this event, like the contents of the packet.
+	Tags []string
+
+	// LoggedPacket is an optional packet metadata.
+	LoggedPacket optional.Value[LoggedPacket]
+}
+
+func newEvent(etype HandshakeEventType, st session.SessionNegotiationState, t time.Time, t0 time.Time) *event {
+	return &event{
+		EventType:    etype,
+		Stage:        st,
+		AtTime:       t.Sub(t0),
+		Tags:         make([]string, 0),
+		LoggedPacket: optional.None[LoggedPacket](),
+	}
+}
+
+// MarshalJSON implements json.Marshaler
+func (e event) MarshalJSON() ([]byte, error) {
+	j := struct {
+		Type   string                       `json:"operation"`
+		Stage  string                       `json:"stage"`
+		Time   float64                      `json:"t"`
+		Tags   []string                     `json:"tags"`
+		Packet optional.Value[LoggedPacket] `json:"packet"`
+	}{
+		Type:   e.EventType.String(),
+		Stage:  e.Stage.String()[2:],
+		Time:   e.AtTime.Seconds(),
+		Tags:   e.Tags,
+		Packet: e.LoggedPacket,
+	}
+	return json.Marshal(j)
+}
+
+// Tracer implements [model.HandshakeTracer].
+type Tracer struct {
+	// events is the array of handshake events.
+	events []*event
+
+	// mu guards access to the events.
+	mu sync.Mutex
+
+	// zeroTime is the time when we started a packet trace.
+	zeroTime time.Time
+}
+
+// NewTracer returns a Tracer with the passed start time.
+func NewTracer(start time.Time) *Tracer {
+	return &Tracer{
+		zeroTime: start,
+	}
+}
+
+// TimeNow allows to manipulate time for deterministic tests.
+func (t *Tracer) TimeNow() time.Time {
+	return time.Now()
+}
+
+// OnStateChange is called for each transition in the state machine.
+func (t *Tracer) OnStateChange(state int) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	stg := session.SessionNegotiationState(state)
```

Comment:
```suggestion

	stg := session.SessionNegotiationState(state)
```

For consistency with functions below.